### PR TITLE
KAFKA-14937: Refactoring for client code to reduce boilerplate

### PR DESCRIPTION
**Pull request by kirktrue**
Move common code from the client implementations to the `ClientUtils` class or the new `ConsumerUtils` class, where possible.

There are a number of places in the client code where the same basic calls are made by more than one client implementation. Minor refactoring will reduce the amount of boilerplate code necessary for the client to construct its internal state.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by philipnee** [comment]
Thanks @kirktrue - I think this makes the existing code much cleaner.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java [hunk 1]
```diff
@@ -17,13 +17,175 @@
...
+    public static final String CONSUMER_METRIC_GROUP_PREFIX = "consumer";
+
+    /**
+     * A fixed, large enough value will suffice for max.
+     */
+    public static final int CONSUMER_MAX_INFLIGHT_REQUESTS_PER_CONNECTION = 100;
+
+    private static final String CONSUMER_CLIENT_ID_METRIC_TAG = "client-id";
+
+    public static ConsumerNetworkClient createConsumerNetworkClient(ConsumerConfig config,
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java]: We are adding all these consumer-related utility functions to this Utils class, that was previously used only for partition comparators, needed for assignment/group coordination. Any reason why we want the new logic here? Maybe better in a separate class?

**Comment by lianetm** [comment]
LGTM, just left one comment above. Thanks for the changes @kirktrue!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java [hunk 2]
```diff
@@ -17,13 +17,175 @@
...
+                netClient,
+                metadata,
+                time,
+                retryBackoffMs,
+                config.getInt(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG),
+                heartbeatIntervalMs);
+    }
+
+    public static LogContext createLogContext(ConsumerConfig config, GroupRebalanceConfig groupRebalanceConfig) {
+        String groupId = String.valueOf(groupRebalanceConfig.groupId);
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java]: The original code has the logic to handle `groupRebalanceConfig.groupId` being null. Should we keep that?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java [hunk 3]
```diff
@@ -17,13 +17,175 @@
...
+        return new Metrics(metricConfig, reporters, time, metricsContext);
+    }
+
+    public static FetchMetricsManager createFetchMetricsManager(Metrics metrics) {
+        Set<String> metricsTags = Collections.singleton(CONSUMER_CLIENT_ID_METRIC_TAG);
+        FetchMetricsRegistry metricsRegistry = new FetchMetricsRegistry(metricsTags, CONSUMER_METRIC_GROUP_PREFIX);
+        return new FetchMetricsManager(metrics, metricsRegistry);
+    }
+
+    public static FetchConfig<String, String> createFetchConfig(ConsumerConfig config) {
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java]: This seems to be only imported, but not used?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java [hunk 4]
```diff
@@ -102,7 +109,10 @@ public DefaultBackgroundThread(final Time time,
                                    final BlockingQueue<ApplicationEvent> applicationEventQueue,
                                    final BlockingQueue<BackgroundEvent> backgroundEventQueue,
                                    final ConsumerMetadata metadata,
-                                   final KafkaClient networkClient) {
+                                   final ApiVersions apiVersions,
+                                   final Metrics metrics,
+                                   final Sensor fetcherThrottleTimeSensor,
+                                   final SubscriptionState subscriptions) {
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java]: subscriptions seems unused?

**Review by junrao — COMMENTED** [review]
@kirktrue : Thanks for the PR. Left a few comments. Also, do you know why there are 94 test failures?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java (hunk 1 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java]: I created a new class named `ConsumerUtils` to put all the new utility methods, leaving `Utils` alone.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java (hunk 3 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java]: Removed.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java (hunk 4 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java]: Removed.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java (hunk 2 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java]: This is an example of the original code from `KafkaConsumer`:

```java
this.groupId = Optional.ofNullable(groupRebalanceConfig.groupId);
this.clientId = config.getString(CommonClientConfigs.CLIENT_ID_CONFIG);

LogContext logContext;

// If group.instance.id is set, we will append it to the log context.
if (groupRebalanceConfig.groupInstanceId.isPresent()) {
    logContext = new LogContext("[Consumer instanceId=" + groupRebalanceConfig.groupInstanceId.get() +
            ", clientId=" + clientId + ", groupId=" + groupId.orElse("null") + "] ");
} else {
    logContext = new LogContext("[Consumer clientId=" + clientId + ", groupId=" + groupId.orElse("null") + "] ");
}
```

In the original code, the `groupRebalanceConfig.groupId` String is converted to an `Optional<String>`. If the `String` was `null`, the `Optional<String>` will return `null`. In that way, the `String` passed to the `LogContext` constructor will include `groupId=null`.

Here's the new code, wrapped up in the `createLogContext` utility method:

```java
public static LogContext createLogContext(ConsumerConfig config, GroupRebalanceConfig groupRebalanceConfig) {
    String groupId = String.valueOf(groupRebalanceConfig.groupId);
    String clientId = config.getString(CommonClientConfigs.CLIENT_ID_CONFIG);
    String logPrefix;
    String groupInstanceId = groupRebalanceConfig.groupInstanceId.orElse(null);
    
    if (groupInstanceId != null) {
        // If group.instance.id is set, we will append it to the log context.
        logPrefix = String.format("[Consumer instanceId=%s, clientId=%s, groupId=%s] ", groupInstanceId, clientId, groupId);
    } else {
        logPrefix = String.format("[Consumer clientId=%s, groupId=%s] ", clientId, groupId);
    }
    
    return new LogContext(logPrefix);
}
```

In the new code, we don't need to convert `groupRebalanceConfig.groupId` into an `Optional<String>`, we just leave it as a plain old `String`. If `groupRebalanceConfig.groupId` is `null`, the result of `logPrefix` will include  `groupId=null`, as before.

I guess we could simplify it even more:

```java
public static LogContext createLogContext(ConsumerConfig config, GroupRebalanceConfig groupRebalanceConfig) {
    String clientId = config.getString(CommonClientConfigs.CLIENT_ID_CONFIG);
    String logPrefix;
    String groupInstanceId = groupRebalanceConfig.groupInstanceId.orElse(null);

    if (groupInstanceId != null) {
        // If group.instance.id is set, we will append it to the log context.
        logPrefix = String.format("[Consumer instanceId=%s, clientId=%s, groupId=%s] ", groupInstanceId, clientId, groupRebalanceConfig.groupId);
    } else {
        logPrefix = String.format("[Consumer clientId=%s, groupId=%s] ", clientId, groupRebalanceConfig.groupId);
    }

    return new LogContext(logPrefix);
}
```

That is, simply remove the first line of the utility method and pass `groupRebalanceConfig.groupId` to the `String.format` call. Internally, `format` will check for any `null` arguments, so the end result would still be `groupId=null`.

**Comment by junrao** [comment]
@kirktrue : Thanks for the updated PR. It LGTM. However, it seems there are built failures?

**Comment by kirktrue** [comment]
@junrao The tests ran again with (no code change in between) and now there are only four test failures in (seemingly) unrelated areas of the code.

**Comment by junrao** [comment]
@kirktrue : It seems there were 4 test failures for jdk 11. But the tests for jdk 17 and 20 were aborted. Do you know why?

**Comment by kirktrue** [comment]
> @kirktrue : It seems there were 4 test failures for jdk 11. But the tests for jdk 17 and 20 were aborted. Do you know why?

@junrao I don't know why yet, no 😞

I have performed some ad hoc testing on my local box (M1) against JDKs 8, 11, 17, and 20 of a handful of the tests that failed. As expected, "it works on my machine," and I get no failures. I have not yet run the whole test suite against each of these JDKs locally. I can do that, but it seems like a brute force approach and I'm not sure it will reveal anything 🤔

Do you have pointers to how to debug test execution in the Apache Jenkins environment?

**Comment by junrao** [comment]
@kirktrue : In the latest run, it seems that most test failures are from java 11. I'd suggest looking at the jenkins's log to see  the first test that failed. Sometimes, a failed test could lead to other test failures, if not designed well.

**Comment by kirktrue** [comment]
@junrao In [the JDK 11 run](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-13990/23/tests), `DeleteTopicTest.testDeleteTopicWithCleaner()` fails and the subsequent integration tests on the same `Executor` fail immediately because of the `verifyNoUnexpectedThreads` check:

```
org.opentest4j.AssertionFailedError: Found 1 unexpected threads during @AfterAll: `Test worker-EventThread` ==> expected: <true> but was: <false>
```

`DeleteTopicTest.testDeleteTopicWithCleaner()` fails during test cleanup:

```
java.nio.file.NoSuchFileException: /tmp/kafka-17102487506151712666/test-0.60ecffe3f71048b49fa8da268bdc7d17-delete/00000000000000000038.snapshot
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
	at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
	at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:149)
	at java.base/sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:99)
	at java.base/java.nio.file.Files.readAttributes(Files.java:1764)
	at java.base/java.nio.file.FileTreeWalker.getAttributes(FileTreeWalker.java:219)
	at java.base/java.nio.file.FileTreeWalker.visit(FileTreeWalker.java:276)
	at java.base/java.nio.file.FileTreeWalker.next(FileTreeWalker.java:373)
	at java.base/java.nio.file.Files.walkFileTree(Files.java:2761)
	at java.base/java.nio.file.Files.walkFileTree(Files.java:2797)
	at org.apache.kafka.common.utils.Utils.delete(Utils.java:899)
	at kafka.utils.CoreUtils$.$anonfun$delete$1(CoreUtils.scala:80)
	at kafka.utils.CoreUtils$.$anonfun$delete$1$adapted(CoreUtils.scala:80)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:933)
	at kafka.utils.CoreUtils$.delete(CoreUtils.scala:80)
	at kafka.utils.TestUtils$.$anonfun$shutdownServers$2(TestUtils.scala:275)
	. . .
```

I am still digging into the details, so I'm not sure what the proper "fix" is, yet.

**Comment by kirktrue** [comment]
@junrao The JDK 11, 17, and 20 test runs output this as their last test execution:

```
Gradle Test Run :streams:integrationTest > Gradle Test Executor 177 > StoreQueryIntegrationTest > shouldQuerySpecificStalePartitionStoresMultiStreamThreadsNamedTopology() PASSED
streams-3: SMOKE-TEST-CLIENT-CLOSED
streams-7: SMOKE-TEST-CLIENT-CLOSED 
streams-5: SMOKE-TEST-CLIENT-CLOSED
streams-6: SMOKE-TEST-CLIENT-CLOSED
streams-0: SMOKE-TEST-CLIENT-CLOSED
streams-8: SMOKE-TEST-CLIENT-CLOSED
streams-1: SMOKE-TEST-CLIENT-CLOSED
streams-2: SMOKE-TEST-CLIENT-CLOSED
streams-4: SMOKE-TEST-CLIENT-CLOSED
streams-2: SMOKE-TEST-CLIENT-CLOSED
streams-5: SMOKE-TEST-CLIENT-CLOSED
streams-0: SMOKE-TEST-CLIENT-CLOSED
streams-1: SMOKE-TEST-CLIENT-CLOSED
streams-4: SMOKE-TEST-CLIENT-CLOSED
streams-3: SMOKE-TEST-CLIENT-CLOSED
```

The JDK 11 test run follows that test with the following last lines:

```
Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
        
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
        
For more on this, please refer to https://docs.gradle.org/8.2.1/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
        
BUILD SUCCESSFUL in 3h 19m 29s
231 actionable tasks: 124 executed, 107 up-to-date
        
Publishing build scan...
https://ge.apache.org/s/k4bsfdqsiuqow
        
        
See the profiling report at: file:///home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-13990/build/reports/profile/profile-2023-07-21-00-27-33.html
A fine-grained performance profile is available: use the --scan option.
Recording test results
[Checks API] No suitable checks publisher found.
Skipping Kafka Streams archetype test for Java 11
```

The JDK 17 test run conclusion is similar, but there's some token publishing error or something:

```
Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.2.1/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 1h 56m 19s
231 actionable tasks: 124 executed, 107 up-to-date

Publishing build scan...

Publishing failed.

The response from https://ge.apache.org/scans/publish/gradle/3.13.4/token was not from Gradle Enterprise.
The specified server address may be incorrect, or your network environment may be interfering.

Please report this problem to your Gradle Enterprise administrator via https://ge.apache.org/help and include the following via copy/paste:

----------
Gradle version: 8.2.1
Plugin version: 3.13.4
Request URL: https://ge.apache.org/scans/publish/gradle/3.13.4/token
Request ID: 97dd78d0-adac-4a90-8f1b-f41b63e6a492
Response status code: 502
Response content type: text/html
----------


See the profiling report at: file:///home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-13990/build/reports/profile/profile-2023-07-21-00-24-49.html
A fine-grained performance profile is available: use the --scan option.
Recording test results
[Checks API] No suitable checks publisher found.
Skipping Kafka Streams archetype test for Java 17
```

However, JDK 20 ends with a hard error due to a timeout or something:

```
Sending interrupt signal to process

> Task :clients:unitTest FAILED
The message received from the daemon indicates that the daemon has disappeared.
Build request sent: Build{id=4413ed42-5d9b-4fcd-94bf-48aaeb0f3abc, currentDir=/home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-13990@2}
Attempting to read last messages from the daemon log...
script returned exit code 143
```

I don't have knowledge of the build process and Jenkins enough to know _why_ this is happening. I'll keep digging.

**Comment by junrao** [comment]
@kirktrue : For JDK 11 tests, it seems that there were 2 checkstyle failures `:clients:checkstyleTest`and `:server-common:checkstyleTest`.

**Comment by kirktrue** [comment]
@junrao in the logs for the JDK 11 run, I see this:

```
> Task :clients:testClasses
Unexpected exception thrown.
org.gradle.internal.remote.internal.MessageIOException: Could not read message from '/127.0.0.1:53070'.
	at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:94)
	at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:270)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException
	at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:72)
	at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:52)
	at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:81)
	... 6 more
> Task :clients:spotbugsTest SKIPPED
> Task :connect:mirror-client:compileTestJava
> Task :connect:mirror-client:testClasses
> Task :storage:api:compileTestJava
> Task :storage:api:testClasses
> Task :log4j-appender:compileTestJava
> Task :log4j-appender:testClasses
org.gradle.internal.remote.internal.ConnectException: Could not connect to server [871f8055-03e7-4c5c-bc92-6b367d765631 port:37149, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1].
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:103)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:774)
	at java.base/sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:120)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54)
	... 5 more
> Task :streams:examples:compileTestJava
> Task :connect:mirror-client:spotbugsTest SKIPPED
> Task :storage:api:spotbugsTest SKIPPED
> Task :streams:examples:testClasses
> Task :clients:checkstyleTest FAILED
```

and a bit later...

```
> Task :server-common:spotbugsTest SKIPPED
Unexpected exception thrown.
org.gradle.internal.remote.internal.MessageIOException: Could not read message from '/127.0.0.1:59256'.
	at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:94)
	at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionReceive.run(MessageHub.java:270)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException
	at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:72)
	at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageReader.read(InterHubMessageSerializer.java:52)
	at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:81)
	... 6 more
org.gradle.internal.remote.internal.ConnectException: Could not connect to server [e16cf35b-4904-4d92-9fc6-98fcebee65e3 port:46581, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1].
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:103)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:774)
	at java.base/sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:120)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54)
	... 5 more
> Task :server-common:checkstyleTest FAILED
```

**Comment by junrao** [comment]
@kirktrue : It seems that JDK 20 tests didn't complete because of failures in ':storage:unitTest'. It would be useful to verify if this is an existing issue.

**Comment by kirktrue** [comment]
@junrao:

> @kirktrue : It seems that JDK 20 tests didn't complete because of failures in ':storage:unitTest'. It would be useful to verify if this is an existing issue.

Looking at the logs for that test run, there are no individual test failures shown in the logs or in the Jenkins UI for the storage module's unit tests.

According to the Gradle Enterprise tool I recently found out about, there were [~200 cases in the last 28 days where test runs failed with that error](https://ge.apache.org/scans/failures?failures.expandedFailures=WzEsMl0&search.relativeStartTime=P28D&search.rootProjectNames=kafka&search.timeZoneId=America/Los_Angeles):

```
Execution failed for task *.
> Process 'Gradle Test Executor *' finished with non-zero exit value 1
  This problem might be caused by incorrect test process configuration.
  For more on test execution, please refer to https://docs.gradle.org/<Gradle version>/userguide/java_testing.html#sec:test_execution in the Gradle documentation.
```

**Comment by junrao** [comment]
Thanks @kirktrue. @satishd : Are you aware of any transient failures in :storage:unitTest? Thanks.

**Comment by kirktrue** [comment]
Here are the details for [build 27](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13990/27/):

* `CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()` is a known flaky test, with [KAFKA-8115](https://issues.apache.org/jira/browse/KAFKA-8115) failed quite a while back to address it.
* `MirrorConnectorsIntegrationExactlyOnceTest.testOffsetTranslationBehindReplicationFlow()` is also a known flaky test. [KAFKA-15197](https://issues.apache.org/jira/browse/KAFKA-15197) was filed because [it fails/flakes nearly half the time it is run](https://ge.apache.org/scans/tests?search.relativeStartTime=P28D&search.rootProjectNames=kafka&search.timeZoneId=Europe/Berlin&tests.container=org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest&tests.test=testOffsetTranslationBehindReplicationFlow()).
* `ZkMigrationIntegrationTest.testDualWriteQuotaAndScram()` is not nearly as flaky and only showed up in this last run. According to [Gradle Enterprise](https://ge.apache.org/scans/tests?search.relativeStartTime=P28D&search.rootProjectNames=kafka&search.timeZoneId=Europe/Berlin&tests.container=kafka.zk.ZkMigrationIntegrationTest&tests.test=testDualWriteQuotaAndScram(ClusterInstance)%5B1%5D), it fails less than 1% of the time.

**Comment by kirktrue** [comment]
> @satishd : Are you aware of any transient failures in :storage:unitTest? Thanks.

According to [Gradle Enterprise](https://ge.apache.org/scans/failures?failures.failureMessage=Execution%20failed%20for%20task%20%27:storage:unitTest%27.%0A%3E%20Process%20%27Gradle%20Test%20Executor%20*%20finished%20with%20non-zero%20exit%20value%201%0A%20%20This%20problem%20might%20be%20caused%20by%20incorrect%20test%20process%20configuration.%0A%20%20For%20more%20on%20test%20execution%2C%20please%20refer%20to%20https://docs.gradle.org/8.2.1/userguide/java_testing.html%23sec:test_execution%20in%20the%20Gradle%20documentation.&search.relativeStartTime=P28D&search.rootProjectNames=kafka&search.timeZoneId=America/Los_Angeles), it's failed 71 times starting on July 18, 2023.

**Comment by satishd** [comment]
@junrao I was not aware of transient failures in storage module. Filed https://issues.apache.org/jira/browse/KAFKA-15285 to track it.

**Comment by kirktrue** [comment]
@junrao In my findings, the issues with the Jenkins runs have either been due to intermittent infrastructure issues, or caused by tests known to be flaky. Suggestions on how to proceed with this PR are welcomed. Thanks!

**Comment by junrao** [comment]
@kirktrue : For consistent infra issues, we could file an infra jira for the ASF infra team to help. For transient test failures, if we could identify when the test failure first occurred and correlate it with the PR that might have introduced it, we could file a jira and ping the PR author to take a look.

**Comment by kirktrue** [comment]
@junrao It doesn't seem that our [test history](https://ge.apache.org/scans/tests?search.relativeStartTime=P90D&search.rootProjectNames=kafka&search.timeZoneId=America/Los_Angeles) goes back much further than a few weeks, so it's hard to see when the more egregious flaky tests crept in.

**Comment by kirktrue** [comment]
@junrao I created this INFRA Jira just now to see if they can help resolve some of the build issues: https://issues.apache.org/jira/browse/INFRA-24874

**Comment by kirktrue** [comment]
> @kirktrue : For consistent infra issues, we could file an infra jira for the ASF infra team to help. For transient test failures, if we could identify when the test failure first occurred and correlate it with the PR that might have introduced it, we could file a jira and ping the PR author to take a look.

@jun done: https://issues.apache.org/jira/browse/INFRA-24874

**Comment by kirktrue** [comment]
@junrao I've rebased against `trunk` over the last week to trigger Jenkins runs. Each build fails with a different set of tests. Most of the failing tests already have Jiras filed.

I do want to point out another pull request I made (#14095) in which I literally only changed `README.md`. The first change was to add a newline, and that build included six test failures. I then removed the newline, and the resulting build included a different set of eight test failures.

According to the graphs on [this Grade Enterprise page](https://ge.apache.org/scans/tests?search.names=Git%20branch&search.relativeStartTime=P28D&search.rootProjectNames=kafka&search.timeZoneId=America/Los_Angeles&search.values=trunk&tests.help=advancedQuery), for the last 28 days Jenkins builds are failing on `trunk` for 28% of the builds and 90% of the builds include flaky tests.

**Comment by junrao** [comment]
@kirktrue : Thanks for the update. If there is clear evidence that test failures are unrelated, it's reasonable to merge the PR. The latest build has 3 red failures due to a connect checkstyle issue tough.

**Comment by junrao** [comment]
@kirktrue : JDK 17 build was aborted. Did that pass locally?

**Comment by philipnee** [comment]
Hey @junrao - from the previous runs.  it seems like JDK17 has been completing with flaky tests. https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/activity?branch=PR-13990

For example: these are the unrelated flaky tests from build #36 (https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-13990/36/tests).
```

Build / JDK 17 and Scala 2.13 / testOffsetTranslationBehindReplicationFlow() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest
2m 27s
Build / JDK 17 and Scala 2.13 / testSyncTopicConfigs() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest
3m 45s
Build / JDK 17 and Scala 2.13 / testSyncTopicConfigs() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationSSLTest
1m 23s
Build / JDK 17 and Scala 2.13 / testRestartReplication() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationSSLTest
1m 8s
Build / JDK 17 and Scala 2.13 / testThrottledProducerConsumer(String).quorum=zk – kafka.api.ClientIdQuotaTest
7s
Build / JDK 17 and Scala 2.13 / testClose() – kafka.api.ConsumerBounceTest
13s
Build / JDK 17 and Scala 2.13 / testBalancePartitionLeaders() – org.apache.kafka.controller.QuorumControllerTest
13s
Build / JDK 17 and Scala 2.13 / testBalancePartitionLeaders() – org.apache.kafka.controller.QuorumControllerTest
15s
Build / JDK 17 and Scala 2.13 / testTaskRequestWithOldStartMsGetsUpdated() – org.apache.kafka.trogdor.coordinator.CoordinatorTest
```

**Review by junrao — APPROVED** [review]
@philipnee Thanks for the explanation. The PR LGTM then.

**Pull request merged**